### PR TITLE
[radiobrowser] Initial contribution

### PR DESCRIPTION
**Pull request by Skinah**
A new binding to navigate and find internet radio broadcasts.
This functionality has been built in to Home Assistant for > 2 years, so it scores high on how many systems use it, as it's installed by default and they need to opt out. 71.6% of their systems at this point in time due to this.
So I feel its time we have the ability for users to come across and have similar ability, done in a different and hopefully better way that has actually helpful documentation.

I feel this is far more flexible than HA's current implementation, which is funny as it is now 2 years on and basic stuff did not get added that users are asking for.
In this binding you can apply multiple filters at a time and can automate the filter changes as they are channels. You can filter to USA + Texas + Spanish + Polka Music to probably get no results with that combo, but its flexible to turn the > 45,000 radio stations down to a short list to try streaming by applying multiple filters.

I added a widget to make using it easier, as a URL text string is not useful for anyone.
https://community.openhab.org/t/radio-browser-basic-widget-for-finding-internet-radio-streams-with-the-ui/153783

### CODE REVIEW

File: bundles/org.openhab.binding.radiobrowser/README.md
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/README.md:14]: Can you fix the table markup?

File: bundles/org.openhab.binding.radiobrowser/README.md
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/README.md:24]: I know it is an advanced channel, but for users it would also be simplify this configuration like `key1=value1,key2=value2` and validate/construct an uri parameter string from that.

File: bundles/org.openhab.binding.radiobrowser/README.md
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/README.md:20]:
```suggestion
The `advanced` configuration parameter `filters` can be used to limit the stations based on different fields like codecs, minimum quality, ordering and more.
All possible filter options are listed here <https://de1.api.radio-browser.info/#Advanced_station_search>
```

File: bundles/org.openhab.binding.radiobrowser/README.md
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/README.md:21]:
```suggestion
The default filter (below) can be changed to suit your needs.
```

File: bundles/org.openhab.binding.radiobrowser/README.md
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/README.md:34]:
```suggestion
```

File: bundles/org.openhab.binding.radiobrowser/README.md
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/README.md:92]: While i wish you a lot of donations and coffee, I'm not sure if this is allowed to be added to the official documentation of the openhab website. Need some advice from other @openhab/add-ons-maintainers if this is oke or not.

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserBindingConstants.java
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserBindingConstants.java:27]: Is there a specific reason to keep the timeout that long? I would say a timeout for 15 seconds is allready very long.

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserBindingConstants.java
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserBindingConstants.java:42]:
```suggestion
    public static final String CHANNEL_ADD_FAVORITE = "add-favorite";
```
Please also change all related occurences.
Documentation: https://www.openhab.org/docs/developer/guidelines.html#naming-convention

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserBindingConstants.java
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserBindingConstants.java:43]:
```suggestion
    public static final String CHANNEL_REMOVE_FAVORITE = "remove-favorite";
```
Please also change all related occurences.

File: bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/binding/binding.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/binding/binding.xml:1]: Remove this 3.x file.

File: bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml:39]: It might be me, as English is not my native language, but i cant understand what this means.

File: bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml:66]:
```suggestion
		<description>A list of stations that you can select from</description>
```

Last part seems technicall and is of no use to the end-user, i guess.

File: bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml:108]:
```suggestion
		<label>Stream URI</label>
```

(does this make sense?)

File: bundles/pom.xml
Comment (lsiepel) [inline bundles/pom.xml:5]: Somehow this PR introdeuced many whitelines, can you revert and only add this binding.

File: bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/addon/addon.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/addon/addon.xml:9]: Can you add the `connection` tag

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserHandler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserHandler.java:76]:
```suggestion
            } else if (command instanceof StringType commandAsString) {
```
Minor improvement regarding java 17 pattern matching. 
If this is change,d all the `command.toString()` can be changed to `commandAsString`

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserHandler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserHandler.java:94]:
```suggestion
```

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserHandler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserHandler.java:97]:
```suggestion
```

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/api/RadioBrowserApi.java
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/api/RadioBrowserApi.java:84]:
```suggestion
        logger.debug("Sending GET:{}", url);
```

If needed, i would suggest the request/response contents to trace, and the rest to debug.

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/api/RadioBrowserApi.java
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/api/RadioBrowserApi.java:78]: Is this available in https ? If so, please adjust.

**Review by lsiepel — CHANGES_REQUESTED** [review]
Looks like a very nice binding. Just finished my first pass of review, checked all files.  There are some questions and suggestions, let me knwo if something is unclear.

File: bundles/org.openhab.binding.radiobrowser/README.md
Comment (Skinah) [reply bundles/org.openhab.binding.radiobrowser/README.md:14]: done

File: bundles/org.openhab.binding.radiobrowser/README.md
Comment (Skinah) [reply bundles/org.openhab.binding.radiobrowser/README.md:24]: I agree it would make it slightly more simple so have done it, but do not believe it was worth the work changing it and then redoing documentation that is already written for the small gain.

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserBindingConstants.java
Comment (Skinah) [reply bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserBindingConstants.java:27]: When I first started using the API I had it set to 10 seconds and got a timeout. It may have been a server update in progress. I have put it back to 10 and will do some smoke tests after having made all your suggestions.

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserBindingConstants.java
Comment (Skinah) [reply bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserBindingConstants.java:42]: When and Why did that change?
See here it used to be lowerCamelCase
https://v34.openhab.org/docs/developer/guidelines.html#java-coding-style
I always make it easy to change this by using the constants, however this is a huge pain to someone that has a binding waiting for a review for 12+ months and then gets hit with this then never wants to make another binding.
Since I do help with the odd review, how do we know this change has been made? Announcements??
Sorry wrong place I know.

File: bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Skinah) [reply bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml:39]: The languages are pulled from their server and contain many typos as it is added by public end users. By using this we throw away some real languages that are not common to get rid of the typos.

English
Englsih
Englihs
American English
British Engligh
etc....

The API tells us how many stations each one has, so we can reduce the list down to just the main common languages.

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserHandler.java
Comment (Skinah) [reply bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserHandler.java:78]: I did try that before making the PR, and I tried again now to get the error. Eclipse is saying that the StringType can not be cast to String. Is using commandAsString.toString() better in any way or I am missing something?

File: bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/addon/addon.xml
Comment (Skinah) [reply bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/addon/addon.xml]: Done

File: bundles/org.openhab.binding.radiobrowser/README.md
Comment (Skinah) [reply bundles/org.openhab.binding.radiobrowser/README.md:24]: I do not understand, so will wait for further clarification on this.

File: bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [reply bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml:114]: > NOOOOO. I do not want two word labels. The risk is that widgets break every time someone in the project decides that items should have no spaces Equipment_UpperCamelCase, then it gets moved to "-" spaced Equipment_Upper_Camel_Case and then to who knows what or if the change gets reversed.
> 
> Discussion on that we now have multiple widgets that are broken due to a core change.
> 
> https://community.openhab.org/t/mediaplayer-now-playing-and-control/126621/8
> 
> I have seen this madness go around in circles since I started with openhab 1.x and sadly it seems to not have stopped yet.

I'm sorry, but label rules are allready stable for a very long time. They should be 2-3 words and < 23 characters. The label as one word is perfectly fine, but it isn't a stream , its a uri.

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserBindingConstants.java
Comment (jlaur) [reply bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserBindingConstants.java]: The naming convention was established in April 2023 after [discussions](https://github.com/openhab/openhab-distro/discussions/1504) in the maintainer group - resulting in: openhab/openhab-docs#2061.

File: bundles/org.openhab.binding.radiobrowser/README.md
Comment (lsiepel) [reply bundles/org.openhab.binding.radiobrowser/README.md:24]: Discussion: https://github.com/openhab/openhab-docs/issues/999
```suggestion
```text
```

File: bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [reply bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml:39]:
```suggestion
				<description>The minimum number of stations a language needs to have, before the binding lists it as an
					option.</description>
```

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserHandler.java
Comment (lsiepel) [reply bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/RadioBrowserHandler.java:78]: Thje StringType seems to work different then i expected. Forget my comment.

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/api/RadioBrowserApi.java
Comment (lsiepel) [reply bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/api/RadioBrowserApi.java:78]: It was set to https and reverted. It l.ooks liek the API is available in https, anything i'm missing?

File: bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Skinah) [reply bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml:39]: If this is hard to understand, should it be made ADVANCED? The default works fine, so it's just for those that want a fast way to get a better result without having to resort to METADATA creating state options from scratch.

File: bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Skinah) [reply bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml:114]: I strongly believe that stream is acceptable and does not confuse an end user. The URI points to a STREAM and only a STREAM so for me this makes sense. Otherwise you could say the same about the ICON channel, it holds a URI that points to an ICON.

Lastly I feel it is more consistent between textual users that see it in the UI and get confused. If the Label is the same as the actual channel I find that far less confusing as the channel is the same and you know that the naming convention never changes and is consistent :)

I do not care if it changes, I just feel strongly against 2 and more words for a binding that is going to rely heavily on a widget working out of the box.

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/api/RadioBrowserApi.java
Comment (Skinah) [reply bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/api/RadioBrowserApi.java:78]: Yes my time, leaving this open to look at it when I have the time. I have not done https with Jetty before so do not know what needs to change without looking deeper into it. Simply changing the one line to https meant the server gave back 404 errors, guessing it needs extra code to setup the certificate.

Comment (Skinah) [reply bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/api/RadioBrowserApi.java:78]: Got it working via HTTPS

File: bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [reply bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml:40]: Not that hard to understand, it was just the world wide part that made it (tome) a bit vague. Now that you explained it, it made sense. As that part doesn;t add much, i sugggest to remove those words (see previous comment)

Regarding the question to mark it advanced or not, it is up to you. For both there are some pros and cons. I don't have a specific opinion on this.

File: bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [reply bundles/org.openhab.binding.radiobrowser/src/main/resources/OH-INF/thing/thing-types.xml:120]: Ok. I understand your pov, lets leave it as is.

FTR: for an icon i have the same opinion :-)

File: bundles/org.openhab.binding.radiobrowser/README.md
Comment (Skinah) [reply bundles/org.openhab.binding.radiobrowser/README.md]: Have removed this from the readme and will instead reply on the widget to make this suggestion.

**Comment by Skinah** [comment]
@lsiepel Thank you for the review, I have address all comments now.

File: bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/api/RadioBrowserApi.java
Comment (lsiepel) [inline bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/api/RadioBrowserApi.java:239]: Are you sure? I would have expected the + to already be encoded as %2B

Comment (Skinah) [reply bundles/org.openhab.binding.radiobrowser/src/main/java/org/openhab/binding/radiobrowser/internal/api/RadioBrowserApi.java:239]: Yes I am sure, have used this a lot in other bindings.
https://stackoverflow.com/questions/4737841/urlencoder-not-able-to-translate-space-character

**Review by lsiepel — APPROVED** [review]
Thank you for contributing this new binding. LGTM.
Do you need more time to test, as you are still adding commits?

**Comment by Skinah** [comment]
I cannot find any more bugs, and a small number of users have reported it works on forum, so let's merge.

**Pull request merged**
